File: segundo/ED/examenes/febrero2019/java/src/Kruskal.java

/**----------------------------------------------
 * -- Estructuras de Datos.  2018/19
 * -- 2º Curso del Grado en Ingeniería [Informática | del Software | de Computadores].
 * -- Escuela Técnica Superior de Ingeniería en Informática. UMA
 * --
 * -- Examen 4 de febrero de 2019
 * --
 * -- ALUMNO/NAME: Miguel Angel Dorado Maldonado
 * -- GRADO/STUDIES: Software A
 * -- NÚM. MÁQUINA/MACHINE NUMBER: xxx
 * --
 * ----------------------------------------------
 */

import dataStructures.graph.WeightedGraph;
import dataStructures.graph.WeightedGraph.WeightedEdge;

import dataStructures.dictionary.Dictionary;
import dataStructures.dictionary.HashDictionary;
import dataStructures.priorityQueue.PriorityQueue;
import dataStructures.priorityQueue.LinkedPriorityQueue;
import dataStructures.set.Set;
import dataStructures.set.HashSet;

public class Kruskal {
  public static <V, W> Set<WeightedEdge<V, W>> kruskal(WeightedGraph<V, W> g) {
    // create prority queue
    PriorityQueue<WeightedEdge<V, W>> PQ = new LinkedPriorityQueue<>();
    for (WeightedEdge<V, W> edge : g.edges()) {
      PQ.enqueue(edge);
    }

    // create DICT
    Dictionary<V, V> DICT = new HashDictionary<>();
    for (V vertice : g.vertices()) {
      DICT.insert(vertice, vertice);
    }

    // set resultado
    Set<WeightedEdge<V, W>> T = new HashSet<>();

    while (!PQ.isEmpty()) {
      WeightedEdge<V, W> arista = PQ.first();
      PQ.dequeue();

      if (!representante(arista.source(), DICT).equals(representante(arista.destination(), DICT))) {
        DICT.insert(representante(arista.destination(), DICT), arista.source());
        T.insert(arista);
      }
    }

    return T;
  }

  private static <V> V representante(V v, Dictionary<V, V> DICT) {
    V actual = v;
    while (!actual.equals(DICT.valueOf(actual))) {
      actual = DICT.valueOf(actual);
    }
    return actual;
  }

  // Sólo para evaluación continua / only for part time students
  public static <V, W> Set<Set<WeightedEdge<V, W>>> kruskals(WeightedGraph<V, W> g) {

    // COMPLETAR

    return null;
  }
}

File: segundo/ED/examenes/febrero2019/haskell/Kruskal.hs
----------------------------------------------
-- Estructuras de Datos.  2018/19
-- 2º Curso del Grado en Ingeniería [Informática | del Software | de Computadores].
-- Escuela Técnica Superior de Ingeniería en Informática. UMA
--
-- Examen 4 de febrero de 2019
--
-- ALUMNO/NAME: Miguel Angel Dorado Maldonado
-- GRADO/STUDIES: Software A
-- NÚM. MÁQUINA/MACHINE NUMBER: xxx
--
----------------------------------------------

module Kruskal (kruskal, kruskals) where

import DataStructures.Dictionary.AVLDictionary qualified as D
import DataStructures.Graph.DictionaryWeightedGraph
import DataStructures.PriorityQueue.LinearPriorityQueue qualified as Q

fromJust :: Maybe a -> a
fromJust (Just a) = a

mkPQueue :: (Ord a) => [a] -> Q.PQueue a
mkPQueue xs = foldr Q.enqueue Q.empty xs

llenar xs = mkPQueue xs

kruskal :: (Ord a, Ord w) => WeightedGraph a w -> [WeightedEdge a w]
kruskal g = aux cola first (dicIni (vertices g))
  where
    cola = llenar (edges g)

    dicIni :: (Ord a) => [a] -> D.Dictionary a a
    dicIni [x] = D.insert x x (D.empty)
    dicIni (x : xs) = D.insert x x (dicIni xs)

    first = Q.first cola
    cola' = Q.dequeue cola

    representante :: (Ord a) => a -> D.Dictionary a a -> a
    representante x dic
      | x /= fromJust (D.valueOf x dic) = representante (fromJust (D.valueOf x dic)) dic
      | otherwise = x

    src :: (Ord a, Ord w) => WeightedEdge a w -> a
    src (WE a w b) = a
    dst :: (Ord a, Ord w) => WeightedEdge a w -> a
    dst (WE a w b) = b

    aux :: (Ord a, Ord w) => Q.PQueue (WeightedEdge a w) -> WeightedEdge a w -> D.Dictionary a a -> [WeightedEdge a w]
    aux cola primeroCola dicVertices
      | Q.isEmpty cola && representante (src primeroCola) dicVertices /= representante (dst primeroCola) dicVertices = [primeroCola]
      | Q.isEmpty cola && representante (src primeroCola) dicVertices == representante (dst primeroCola) dicVertices = []
      | not (Q.isEmpty cola) && representante (src primeroCola) dicVertices /= representante (dst primeroCola) dicVertices = primeroCola : aux cola'' first' (D.insert (dst primeroCola) (src primeroCola) dicVertices)
      | not (Q.isEmpty cola) && representante (src primeroCola) dicVertices == representante (dst primeroCola) dicVertices = aux cola'' first' dicVertices
      | otherwise = aux cola'' first' dicVertices
      where
        first' = Q.first cola
        cola'' = Q.dequeue cola

-- Solo para evaluación continua / only for part time students
kruskals :: (Ord a, Ord w) => WeightedGraph a w -> [[WeightedEdge a w]]
kruskals = undefined
